glean-core/src/lib.rs

```diff
@@ -500,6 +514,30 @@ pub fn shutdown() {
         log::error!("Can't shutdown dispatcher thread: {:?}", e);
     }
 
+    // Call on_shutdown.
+    // Need to be done without holding a lock on the global Glean object.
+    {
+        let (tx, rx) = unbounded();
+
+        thread::Builder::new()
+            .name("glean.shutdown".to_string())
+            .spawn(move || {
+                let state = global_state().lock().unwrap();
+                state.callbacks.shutdown();
+
+                // Best-effort sending. The other side might have timed out already.
+                let _ = tx.send(()).ok();
+            })
+            .expect("Unable to spawn thread to wait on shutdown");
+
+        // TODO: 30 seconds? What's a good default here? Should this be configurable?
```

Comment:
For the current situation the shutdown processing duration is dominated by two things: network conditions for the single upload we might try, and whether we're waiting.

Network conditions around upload success and failure are something we can (and maybe should) instrument. We have `glean.upload.ping_upload_failure`, maybe we need a `timing_distribution` `glean.upload.ping_upload_success`? If we're willing to presume that Glean upload is at worst similarly performant to Legacy Firefox Desktop Telemetry upload, [on Beta we're looking at p95 around 4.5s](https://telemetry.mozilla.org/new-pipeline/dist.html#!cumulative=0&end_date=2023-01-29&include_spill=0&keys=__none__!__none__!__none__&max_channel_version=beta%252F110&measure=TELEMETRY_SEND_SUCCESS&min_channel_version=null&processType=*&product=Firefox&sanitize=1&sort_by_value=0&sort_keys=submissions&start_date=2023-01-16&table=0&trim=1&use_submission_date=0) (0.58% of successful uploads are in buckets above 33s). (send failures take longer, but that's likely dominated by timeout statuses).

As for `Wait`, the way I understand it we only `Wait` in two situations: 
1) we're not done processing pending pings, 
2) we're rate-limiting.

In 1 we might `Wait` -> `Wait` -> `Wait` because we don't know how long it takes to process pending pings. There's nothing that'll interrupt the loading of the pending pings, so we might be stuck in `ThreadJoinHandle('glean.upload').join()` for `3 * WAIT_TIME_FOR_PING_PROCESSING = 3s` (I originally worried this'd be `3 * Rate Limiter Interval = 3min`, but processing wisely has its own timing).

In 2 we don't `Wait` -> `Wait` because the rate limiter tells us to wait the full amount of time in one go, and as you say Rust should be trustworthy to get us to the full duration needed for the rate limiter to reset itself and start processing again. In this case we'll go `Wait` -> `Upload` (if there are tasks and we were throttled) meaning we can only get stuck for one full interval (1min). This is suboptimal because the 30s timeout here ensures we'll always wait the full 30s and still not perform any upload at the end of it.

---

In short, I think the 30s is (until we find out differently) a fine interval to choose. I'd like some instrumentation about what situations this new "wait for shutdown" module finds itself in so we can decide to tune it later, but I'm not sure we'll need it to be configurable. Tuning it release-by-release feels good enough.

...and I am a touch bothered by the rate limiter case. If we're in rate limiting then we've processed a bunch of pings which means we're out of danger of the problem this patch aims to solve... but it means we'll be sitting at shutdown for 30s for no benefit whatsoever.

A way out of this could be having us check before we `join` whether we're in `Wait` for longer than the timeout. (We could pass the shutdown timeout like `callbacks.shutdown(timeout)` which could pass it to the uploader which could compare it against the duration we gave to sleep. If it's too long we could opt to not join and allow `glean.shutdown` to fast-exit). It's adding more complication to cover what might be (without instrumentation we'd never know whether it is) a rare occurrence, so I'll not be blocking my review on it.